timemachine/gcmc/mover.py

```diff
@@ -0,0 +1,259 @@
+from timemachine.md import builders
+from timemachine.potentials import nonbonded
+
+CHEMICAL_POTENTIAL = 26.61236542799211  # calibrated using 12 window, staged AHFE
+STANDARD_VOLUME = 0.029884155244114276  # calibrated using end-states of the above
+
+
+def randomly_rotate_and_translate(coords, offset):
+    centroid = np.mean(coords, axis=0, keepdims=True)
+    centered_coords = coords - centroid
+    rotated_coords = np.matmul(centered_coords, special_ortho_group.rvs(3))
+    return rotated_coords + offset
+
+
+class GCMCMover:
+    def __init__(self, on_params, off_params, nb_beta, nb_cutoff):
+        self.on_params = on_params
+        self.off_params = off_params
+        self.nb_beta = nb_beta
+        self.nb_cutoff = nb_cutoff
+        # counters
+        self.i_count = 0
+        self.d_count = 0
+        self.t_count = 0
+
+        @jax.jit
+        def U_fn(conf, box, params, a_idxs, b_idxs):
+            conf_i = conf[a_idxs]
+            conf_j = conf[b_idxs]
+            params_i = params[a_idxs]
+            params_j = params[b_idxs]
+            return nonbonded.nonbonded_block(conf_i, conf_j, box, params_i, params_j, self.nb_beta, self.nb_cutoff)
+
+        self.U_fn = U_fn
+
+    def toggle_gcmc_move(self, coords, box, cur_params, water_idxs, water_states):
+        # this is a little bit different from the Adams formulation since we don't compute the 1/(N+1) or the N
+        # prefactors.
+
+        # 1. randomly turn off/on a water with 50/50 probability
+        #   a. if turning off, compute delta_U when non-interacting
+        #   b. if turning on, randomly rotate and translate, then turn on and compute delta_Us
+
+        # print("num waters", np.sum(water_states))
+
+        # TBD: should we compute cur_params from water_states?
+
+        n_waters = len(water_states)
+        n_atoms = len(coords)
+
+        chosen_water = np.random.randint(0, n_waters)
+        chosen_water_atoms = water_idxs[chosen_water]
+
+        a_idxs = chosen_water_atoms
+        b_idxs = np.setdiff1d(np.arange(n_atoms), chosen_water_atoms)
+
+        self.t_count += 1
+
+        box_vol = np.product(np.diag(box))
+
+        beta = 1 / DEFAULT_KT
+        adams_B = beta * CHEMICAL_POTENTIAL + np.log(box_vol / STANDARD_VOLUME)
+
+        # useful for visualization
+        TRANSLATIONAL_OFFSET = 0
+
+        if water_states[chosen_water]:
+            # turn interacting water off
+
+            trial_params = cur_params.copy()
+
+            np.testing.assert_equal(trial_params[chosen_water_atoms], self.on_params)
+
+            # compute acceptance probability, the - sign is because we're missing
+            # these interactions
+            delta_U = -self.U_fn(coords, box, trial_params, a_idxs, b_idxs)
+            # print("delta_U delete", delta_U)
+            # assert delta_U < 1e4
+
+            # p_delete = min(1, np.exp(-adams_B) * np.exp(-beta * delta_U))
+            p_delete = min(1, np.exp(-adams_B - beta * delta_U))
```

Comment:
Naive question / point of uncertainty: Why is a standard volume required in computing the log acceptance probability?

(For this symmetric `water_states`-toggling proposal, can it not be computed from `reduced_potential(trial_coords, box, trial_water_states) - reduced_potential(coords, box, water_states)`? If so, I see how the definition of `reduced_potential` depends on `CHEMICAL_POTENTIAL` but not how it depends on `STANDARD_VOLUME`.)

Reply:
This is a good point - the standard volume correction is needed in the original GCMC formulation (especially when accounting for the targeted case, i.e. the GCMC region maybe a small sub-region of the full-box).

For the semi-grand symmetric exchange case, insertion into an empty box becomes extremely difficult: the acceptance probability for the first insertion is `exp(-beta*chemical_potential) << 1`. I will re-test this with insertion into a box near equilibrium next.